### ISSUE TITLE
Expose a function to obtain cargo's installation root

### ISSUE THREAD

**Issue by yds12** [rust-lang/cargo#13288]
### Problem

I have an application that needs to know where binaries installed via `cargo install` are located in the system. I can roughly assume it will be in `$HOME/.cargo/bin`, but according to [this](https://doc.rust-lang.org/cargo/commands/cargo-install.html), when we do `cargo install` the location will be determined by, in order of priority:

- --root option
- CARGO_INSTALL_ROOT environment variable
- install.root Cargo [config value](https://doc.rust-lang.org/cargo/reference/config.html)
- CARGO_HOME environment variable
- $HOME/.cargo

My idea was to use `cargo` as a library crate (although I need to see the impact in terms of dependencies) to obtain this location. I assume `cargo` already has some internal function that does that, as far as I checked it uses a `resolve_root` function, but that function is not public.

In summary: it would be nice to have a function in cargo (as a library) that given a current working directory, could give me the installation root.

### Proposed Solution

If I understand correctly, making `resolve_root` public would suffice.

**Comment by epage** [comment]
Could you go into detail on *why* you need to know where the install directory is?  That can help making sure we talk about a solution that is relevant to your problem.  For example, if the user sets `--root`, then I don't think we track that for you.  If they set it via config, then looking it up is dependent on what directory you are in.

**Comment by yds12** [comment]
Hi, thanks for the quick feedback! So, given a certain system-wide available binary `x` I need to be able to tell if `x` was installed via cargo, or not. The reason is that we want to take different actions depending on that. My idea was to compare the output of something like `which x` with cargo's install root to be able to say "yes, it was installed via cargo". I'm doing this in a "best-effort" basis, so it's ok if I miss the case when the user used `--root` or something of that sort.

I just noticed that I can also do `cargo install --list` to get the names of executables installed by cargo (although that would require me to shell -- for context, I'm doing this from rust code). That is enough to solve my problem, so if this feature is too niche, feel free to close the issue.

**Comment by epage** [comment]
So the request isn't "list of binaries cargo installed" but "list of binaries installed into the common location", right?

That reduces your look up to 
- CARGO_HOME environment variable
- $HOME/.cargo

which is covered by [`home::cargo_home`](https://docs.rs/home/latest/home/fn.cargo_home.html).

We might be adding more paths in the future for XDG support (see https://internals.rust-lang.org/t/pre-rfc-split-cargo-home/19747) but that would likely be a more specific entry point in `home` (e.g. `cargo_install_dir`)

**Comment by yds12** [comment]
Shouldn't I also check CARGO_INSTALL_ROOT, in case the user has modified that?

**Comment by epage** [comment]
That depends.  `CARGO_INSTALL_ROOT` is exactly the same as `install.root` and most users who want it more permamently set would likely use `install.root` instwad, so I see `CARGO_INSTALL_ROOT` as more of a transient property which means you likely can't rely on it.
